pkg/api/v1/conversion.go

```diff
@@ -715,51 +689,105 @@ func Convert_v1_ResourceList_To_api_ResourceList(in *v1.ResourceList, out *api.R
 }
 
 func AddFieldLabelConversionsForEvent(scheme *runtime.Scheme) error {
-	return scheme.AddFieldLabelConversionFunc("v1", "Event",
-		func(label, value string) (string, string, error) {
-			switch label {
-			case "involvedObject.kind",
-				"involvedObject.namespace",
-				"involvedObject.name",
-				"involvedObject.uid",
-				"involvedObject.apiVersion",
-				"involvedObject.resourceVersion",
-				"involvedObject.fieldPath",
-				"reason",
-				"source",
-				"type",
-				"metadata.namespace",
-				"metadata.name":
-				return label, value, nil
-			default:
-				return "", "", fmt.Errorf("field label not supported: %s", label)
-			}
-		})
+	return scheme.AddFieldLabelConversionFunc("v1", "Event", EventFieldLabelConversionFunc)
 }
 
 func AddFieldLabelConversionsForNamespace(scheme *runtime.Scheme) error {
-	return scheme.AddFieldLabelConversionFunc("v1", "Namespace",
-		func(label, value string) (string, string, error) {
-			switch label {
-			case "status.phase",
-				"metadata.name":
-				return label, value, nil
-			default:
-				return "", "", fmt.Errorf("field label not supported: %s", label)
-			}
-		})
+	return scheme.AddFieldLabelConversionFunc("v1", "Namespace", NamespaceFieldLabelConversionFunc)
 }
 
 func AddFieldLabelConversionsForSecret(scheme *runtime.Scheme) error {
-	return scheme.AddFieldLabelConversionFunc("v1", "Secret",
-		func(label, value string) (string, string, error) {
-			switch label {
-			case "type",
-				"metadata.namespace",
-				"metadata.name":
-				return label, value, nil
-			default:
-				return "", "", fmt.Errorf("field label not supported: %s", label)
-			}
-		})
+	return scheme.AddFieldLabelConversionFunc("v1", "Secret", SecretFieldLabelConversionFunc)
+}
+
+// FieldLabelConversionFunc converts a field selector to internal representation.
+// Return error for unsupported field selector.
+func GenericFieldLabelConversionFunc(label, value string) (string, string, error) {
+	objectMeta := metav1.ObjectMeta{}
+	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(&objectMeta, true)
+	if objectMetaFieldsSet.Has(label) {
+		return label, value, nil
+	}
+	return "", "", fmt.Errorf("field label %q not supported", label)
+}
+
+// PodFieldLabelConversionFunc converts a field selector to internal representation.
+// Return error for unsupported field selector.
+func PodFieldLabelConversionFunc(label, value string) (string, string, error) {
+	// This is for backwards compatibility with old v1 clients which send spec.host
+	if label == "spec.host" {
+		label = "spec.nodeName"
+	}
+	apiPod := api.Pod{}
+	podFieldSet := PodToSelectableFields(&apiPod)
+	if podFieldSet.Has(label) {
+		return label, value, nil
+	}
+	return "", "", fmt.Errorf("field label %q not supported for %q", label, "Pod")
```

Comment:
Is this 1200 lines of motion being done to provide the name of the kind (not even resource) being used for field selection?  I'm not sure the violence is worth it.  Doesn't a client already know what they're calling?

Reply:
@deads2k These fieldLabel conversion functions have already been registered for `v1`, such as

```
scheme.AddFieldLabelConversionFunc("v1", "Pod", PodFieldLabelConversionFunc)
```

Just like what've done before. I just refactor the codes.